File: src/dialogs/layerpropertiesdialog.cpp
#include "layerpropertiesdialog.h"
#include "ui_layerpropertiesdialog.h"

LayerPropertiesDialog::LayerPropertiesDialog(QWidget *parent): QDialog(parent), ui(new Ui::LayerPropertiesDialog) {
	ui->setupUi(this);
	connect(ui->parallaxChk, &QCheckBox::stateChanged, this, &LayerPropertiesDialog::onParallaxChkStateChanged);
	connect(ui->horizontalParallaxSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onHorizontalParallaxSliderValueChanged);
	connect(ui->verticalParallaxSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onVerticalParallaxSliderValueChanged);
	connect(ui->horizontalAsSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onHorizontalAsSliderValueChanged);
	connect(ui->verticalAsSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onVerticalAsSliderValueChanged);
}

LayerPropertiesDialog::~LayerPropertiesDialog() {
	disconnect(ui->parallaxChk, &QCheckBox::stateChanged, this, &LayerPropertiesDialog::onParallaxChkStateChanged);
	disconnect(ui->horizontalParallaxSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onHorizontalParallaxSliderValueChanged);
	disconnect(ui->verticalParallaxSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onVerticalParallaxSliderValueChanged);
	disconnect(ui->horizontalAsSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onHorizontalAsSliderValueChanged);
	disconnect(ui->verticalAsSlider, &QSlider::valueChanged, this, &LayerPropertiesDialog::onVerticalAsSliderValueChanged);
	delete ui;
}

QString LayerPropertiesDialog::getName() {
	return ui->nameLineEdit->text();
}

void LayerPropertiesDialog::setName(QString name) {
	ui->nameLineEdit->setText(name);
}

int LayerPropertiesDialog::getWidth() {
	return ui->widthSpinBox->value();
}

int LayerPropertiesDialog::getHeight() {
	return ui->heightSpinBox->value();
}

void LayerPropertiesDialog::setSize(int width, int height) {
	ui->widthSpinBox->setValue(width);
	ui->heightSpinBox->setValue(height);
}

int LayerPropertiesDialog::getHorizontalParallax() {
	return ui->horizontalParallaxSlider->value();
}

int LayerPropertiesDialog::getVerticalParallax() {
	return ui->verticalParallaxSlider->value();
}

void LayerPropertiesDialog::setParallax(int h, int v) {
	ui->horizontalParallaxSlider->setValue(h);
	ui->verticalParallaxSlider->setValue(v);
}

bool LayerPropertiesDialog::parallaxEnabled() {
	return ui->parallaxChk->isChecked();
}

void LayerPropertiesDialog::setParallaxEnabled(bool enabled) {
	ui->parallaxChk->setChecked(enabled);
}

void LayerPropertiesDialog::setReflectiveEnabled(bool enabled) {
	ui->reflective_chk->setChecked(enabled);
}

bool LayerPropertiesDialog::reflectiveEnabled() {
	return ui->reflective_chk->isChecked();
}

int LayerPropertiesDialog::getAutoScrollingWidth() {
	return ui->horizontalAsSlider->value();
}

int LayerPropertiesDialog::getAutoScrollingHeight() {
	return ui->verticalAsSlider->value();
}

void LayerPropertiesDialog::setAutoScrolling(int h, int v) {
	ui->horizontalAsSlider->setValue(h);
	ui->verticalAsSlider->setValue(v);
}

void LayerPropertiesDialog::onParallaxChkStateChanged(int arg1) {
	ui->horizontalParallaxSlider->setEnabled(arg1 == Qt::Checked);
	ui->verticalParallaxSlider->setEnabled(arg1 == Qt::Checked);
	ui->horizontalAsSlider->setEnabled(arg1 == Qt::Checked);
	ui->verticalAsSlider->setEnabled(arg1 == Qt::Checked);
}

void LayerPropertiesDialog::onHorizontalParallaxSliderValueChanged(int value) {
	updateSliderIndicator(ui->horizontalParallaxSlider, ui->horizontalParallaxIndicator);
}

void LayerPropertiesDialog::onVerticalParallaxSliderValueChanged(int value) {
	updateSliderIndicator(ui->verticalParallaxSlider, ui->verticalParallaxIndicator);
}

void LayerPropertiesDialog::onHorizontalAsSliderValueChanged(int value) {
	updateSliderIndicator(ui->horizontalAsSlider, ui->horizontalAsIndicator);
}

void LayerPropertiesDialog::onVerticalAsSliderValueChanged(int value) {
	updateSliderIndicator(ui->verticalAsSlider, ui->verticalAsIndicator);
}

void LayerPropertiesDialog::updateSliderIndicator(QSlider* slider, QLabel* indicator) {
	QString padding = "";
	int val = slider->value();
	QString valStr= QString::number(val);

	do {
		padding += " ";
	} while(padding.length() + valStr.length() < 4);
	indicator->setText(padding + valStr);
}

File: dialogs/layerpropertiesdialog.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QDialog, QWidget

from ui.ui_layerpropertiesdialog import Ui_LayerPropertiesDialog

from formats.spheremap import MapLayer, SphereMap

class LayerPropertiesDialog(QDialog):
	ui: Ui_LayerPropertiesDialog


	@property
	def parallaxX(self):
		return self.ui.horizontalParallaxSlider.value() / 10

	@property
	def parallaxY(self):
		return self.ui.verticalParallaxSlider.value() / 10

	@property
	def scrollX(self):
		return self.ui.horizontalAsSlider.value() / 10

	@property
	def scrollY(self):
		return self.ui.verticalAsSlider.value() / 10


	def __init__(self, parent: QWidget = None):
		super().__init__(parent)
		self.ui = Ui_LayerPropertiesDialog()
		self.ui.setupUi(self)
		self.ui.horizontalParallaxSlider.valueChanged.connect(lambda v: self.ui.horizontalParallaxIndicator.setText(f"{v/10}"))
		self.ui.verticalParallaxSlider.valueChanged.connect(lambda v: self.ui.verticalParallaxIndicator.setText(f"{v/10}"))
		self.ui.horizontalAsSlider.valueChanged.connect(lambda v: self.ui.horizontalAsIndicator.setText(f"{v/10}"))
		self.ui.verticalAsSlider.valueChanged.connect(lambda v: self.ui.verticalAsIndicator.setText(f"{v/10}"))


	def __setParallaxScrollControlVals(self, layer:MapLayer):
		self.ui.parallaxChk.setEnabled(True)
		self.ui.parallaxChk.setChecked(layer.hasParallax)
		self.ui.horizontalParallaxSlider.setEnabled(True)
		self.ui.horizontalParallaxSlider.setValue(layer.parallaxX * 10)
		self.ui.horizontalParallaxIndicator.setText(f"{layer.parallaxX}")
		self.ui.verticalParallaxSlider.setEnabled(True)
		self.ui.verticalParallaxSlider.setValue(layer.parallaxY * 10)
		self.ui.verticalParallaxIndicator.setText(f"{layer.parallaxY}")

		self.ui.horizontalAsSlider.setEnabled(True)
		self.ui.horizontalAsSlider.setValue(layer.scrollingX * 10)
		self.ui.horizontalAsIndicator.setText(f"{layer.scrollingX}")
		self.ui.verticalAsSlider.setEnabled(True)
		self.ui.verticalAsSlider.setValue(layer.scrollingY * 10)
		self.ui.verticalAsIndicator.setText(f"{layer.scrollingY}")


	def show(self, layerIndex:int, map:SphereMap):
		layer = map.layers[layerIndex]
		self.ui.nameLineEdit.setText(layer.name)
		self.ui.widthSpinBox.setValue(layer.width)
		self.ui.heightSpinBox.setValue(layer.height)
		self.ui.reflective_chk.setChecked(layer.reflective)

		if layer.hasParallax or layer.scrollingX + layer.scrollingY > 0:
			self.__setParallaxScrollControlVals(layer)
		else:
			for mLayer in map.layers:
				if not mLayer.hasParallax:
					self.__setParallaxScrollControlVals(layer)
		return super().show()
